Batch inference documentation (#103)
odahu/odahu-docs: source/architecture_batch.md

#### 1. Motivation

User wants to run workloads using some trigger (cron, webhook, cloud events) that make predictions on some dataset
using ML Model

This dataset could have relatively large size and using blocking web api does not fit this usecase perfectly

So we should provide async API that provide way to specify 

1. Where input data is located (object storage link: s3, gcs, azure, etc)
2. Where model is located (object storage link: s3, gcs, azure, etc) OR maybe some specific ML Registry such as MLFlow registry
3. Where output data should be uploaded (object storage link: s3, gcs, azure, etc)

> 4. Triggers that allow to run this job

Reviewer: small grammar fixes
``` suggestion
A user wants to run workloads using a trigger (cron, webhook, cloud events) that makes predictions on a dataset using ML Model.

This dataset could have a relatively large size and the use of blocking web API does not fit this use case perfectly.

So we should provide an async API that provides a way to specify: 

1. Where input data is located (object storage link: s3, gcs, azure, etc)
2. Where the model is located (object storage link: s3, gcs, azure, etc) OR maybe some specific ML Registry such as MLFlow registry
3. Where output data should be uploaded (object storage link: s3, gcs, azure, etc)
4. Triggers that allow running this job.
```

#### 2. REST API

New manifests (REST API entities) are introduced

1. BatchPredictor

```yaml
ID: wpm-matching-3-7
Kind: BatchPredictor
Spec:
  image: <link to docker image>
  entrypoint: python
  cmd: ["run.py", "forecast", "-m", "top_emp"]
  inputConnection: input-data
  outputConnection: output-data
  modelConnection: model-data
  # If data will be not available in the moment of launch the predictor 
  # will wait some time (retries * delayBetweenRetries) before stop with exception
  waitForInputData:
    retries: 15  
    delayBetweenRetries: 120 
  triggers:
    webhook:

>       enabled: true

Reviewer: We plan, for now, to add only `webhook` triggers or, maybe, it worth creating a separate entity for triggers? To have the possibility update only trigger URI or type of trigger?

    machine_type: n1-standard-4
  resources: 
    cpu: 2
    memory: 2Gi
Status:
  triggers:
    webhook: <uri> for trigger job
```

2. BatchPredictionJob

```yaml
ID: <generated ID>
Kind: BatchPredictionJob
Spec:
  entrypoint: python  # has priority under the same property in BatchPredictor
  cmd: ["run.py", "forecast", "-m", "top_emp"] # has priority under the same property in BatchPredictor
  modelPath: /model_output/921
  inputPath: /data/2020-20-20
  outputData: /data/2020-20-20
  nodeSelector:  # has priority under the same property in BatchPredictor
    machine_type: n1-standard-4  
  resources:  # has priority under the same property in BatchPredictor
    cpu: 2
    memory: 2Gi
  waitForInputData:  # has priority under the same property in BatchPredictor
    retries: 15
    delayBetweenRetries: 120
Status:
  triggerDate: 2020-02-02 12:30:31
  state: Completed
```

- Spec.modelPath must be directory in remote storage (not file) and must contain
  `odahuflow.model.yaml` file which should have next properties
  ```yaml
  modelName: <Model name>
  modelVersion: 1.3
  ```
  and any other files, for example model configuration that can be loaded to model in inference script
  

> - Spec.inputPath must be a file in remote storage (not directory) and be one of next two options:

Reviewer: ```suggestion
- Spec.inputPath must be a file in remote storage (not a directory) and be one of the next two options:
```

  - option 1: json file that contains data saved in `Inference Request` object 
    according to [Predict Protocol - Version 2](https://github.com/kubeflow/kfserving/blob/master/docs/predict-api/v2/required_api.md#inference-request-json-object)

>   - option 2: .tar.gz file that contain a compressed bundle of json files the same format as json from option 1

Reviewer: ```suggestion
  - option 2: .tar.gz file that contains a compressed bundle of json files in the same format as json from option 1
```

> - Spec.outputPath must be a file in remote storage (not directory) and be one of next two options:

Reviewer: ```suggestion
- Spec.outputPath must be a file in remote storage (not a directory) and be one of the next two options:
```

  - option 1: json file that contains data saved in `Inference Response` object 
    according to [Predict Protocol - Version 2](https://github.com/kubeflow/kfserving/blob/master/docs/predict-api/v2/required_api.md#inference-response-json-object)

>   - option 2: .tar.gz file that contain a compressed bundle of json files the same format as json from option 1

Reviewer: ```suggestion
  - option 2: .tar.gz file that contain a compressed bundle of json files in the same format as json from option 1
```
